server/routes/visitors.ts

```diff
@@ -137,6 +141,70 @@ export default function routes(
     }
   )
 
+  router.post(
+    '/select-date-and-time/:offenderNo',
+    body('visit-date-and-time').custom((value: string, { req }) => {
+      const slotsList = req.session.slotsList as VisitSlotList
+
+      // check selected slot is in the list that was shown and has available tables
+      const selectedSlot: VisitSlot = Object.values(slotsList)
+        .flat()
+        .reduce((allSlots, slot) => {
+          return allSlots.concat(slot.slots.morning, slot.slots.afternoon)
+        }, [])
+        .find(slot => slot.id === value)
+
+      if (selectedSlot === undefined || selectedSlot.availableTables === 0) {
+        throw new Error('No time slot selected')
+      }
+
+      return true
+    }),
+    param('offenderNo').custom((value: string) => {
+      if (!isValidPrisonerNumber(value)) {
+        throw new Error('Invalid prisoner number supplied')
+      }
+
+      return true
+    }),
+    async (req, res) => {
+      const { offenderNo } = req.params
+      const errors = validationResult(req)
+
+      if (!errors.isEmpty()) {
+        return res.render('pages/dateAndTime', {
+          errors: !errors.isEmpty() ? errors.array() : [],
+          offenderNo,
+          prisonerName: req.session.prisonerName,
+          slotsList: req.session.slotsList,
+          timeOfDay: req.session.timeOfDay,
+          dayOfTheWeek: req.session.dayOfTheWeek,
+        })
+      }
+      return res.redirect(`/visit/additional-support/${req.params.offenderNo}`)
+    }
+  )
+
+  router.get(
+    '/additional-support/:offenderNo',
+    param('offenderNo').custom((value: string) => {
```

Comment:
I added this check on `offenderNo` for consistency with the other routes but have realised it doesn't do anything here (and in some other places) because it's not showing errors on `GET` requests.

Have left for time being as thought it worth discussing whether we need to keep the `offenderNo` in the URL all the way through? It's not actually doing anything on some screens; i.e. you could put a different (but valid) one - and be confused! Thinking it might be better to keep it there but use the validation to check it matches the one in the session? (And if it didn't, probably send you back to the beginning as something's gone wrong....)

Reply:
I'd already thought about that tbh, after the initial selection I don't think it's required. It should go in the session 🙂